### PR TITLE
feat(sandbox): Support sshd-based stateful docker session

### PR DESCRIPTION
**Pull request by xingyaoww**
I tried to implement the ssh-based stateful docker sandbox solution @frankxu2004
 [proposed](https://github.com/OpenDevin/OpenDevin/issues/226#issuecomment-2041220938). Turns out it is not hard to get working! 

I haven't implemented `root` login yet, but I think `opendevin` user login works quite well.

See screenshot:
(for some reason, the default logger seems to repeating the same output twice - can try to fix in a separate PR)
image

### CODE REVIEW

**Comment by xingyaoww** [comment]
Fix the logging issue in https://github.com/OpenDevin/OpenDevin/pull/850#issue-2229602109.

Here's the fixed version:

image

**Review by neubig — COMMENTED** [review]
This looks like a nice solution! But just to clarify, this is starting an ssh daemon with a known username and password (opendevin/opendevin). This seems like a potential security hazard -- could people use this to log into your docker container and run code if they found a machine running opendevin?

If that's the case, we should probably do either/both of: 
* Make the sshd only accept connections from localhost (still not perfect, as someone with access to the same computer could cause trouble) 
* make the password a long random string

Separately before merging this I think someone should test to do a sanity check. I'm a bit busy now till Tuesday so hopefully someone else could take a look! If not I'll take a look as soon as I can

**Comment by xingyaoww** [comment]
Thanks for the review @neubig ! That is a very good point! I've added a random user password and set `-o 'ListenAddress=127.0.0.1'` to prevent users outside localhost from connecting to the sandbox.

**Comment by frankxu2004** [comment]
This is great @xingyaoww so fast for implementing this! Thank you!

I think the `pexpect` module can also handle things like prompting to input password, as well as those Y/N prompts in the middle of a process.

Now it's using `ssh.prompt()` to look for shell prompts like `#` or `$` to determine if the previous command is finished as for when to read the output and send the input. 
For processes asking password like "Password:" or apt-get "Proceed? y/n" do we need to add a bit more regex to the default prompt https://pexpect.readthedocs.io/en/stable/api/pexpect.html#pexpect.spawn.expect in order to handle those "wait-for-input", or does the current implementation works out of the box for "y/n" and "password:". I am thinking if `ssh.expect("Password:")` or `ssh.expect("y/n")` is needed

File: opendevin/sandbox/Dockerfile [hunk 1]
```diff
@@ -15,3 +15,6 @@ RUN apt-get update && apt-get install -y \
     python3-dev \
     build-essential \
     && rm -rf /var/lib/apt/lists/*
+
+RUN apt update && apt install openssh-server sudo -y
```
Comment (rbren) [inline opendevin/sandbox/Dockerfile:19]: we should move this up into the `install` list above, right?

**Review by rbren — APPROVED** [review]
Haven't tested locally, but this seems great

File: opendevin/sandbox/Dockerfile (hunk 1 above)
Comment (xingyaoww) [reply opendevin/sandbox/Dockerfile]: good point! done!

**Comment by xingyaoww** [comment]
Great point @frankxu2004 ! Let's start an issue tracking this and we can start a separate PR to get this implemented!

**Pull request merged**